components/web_discovery/browser/background_credential_helper.cc

```diff
@@ -0,0 +1,127 @@
+/* Copyright (c) 2024 The Brave Authors. All rights reserved.
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at https://mozilla.org/MPL/2.0/. */
+
+#include "brave/components/web_discovery/browser/background_credential_helper.h"
+
+#include <utility>
+
+#include "base/base64.h"
+#include "base/containers/span_rust.h"
+#include "base/logging.h"
+#include "base/threading/thread_restrictions.h"
+#include "crypto/sha2.h"
+
+namespace web_discovery {
+
+BackgroundCredentialHelper::BackgroundCredentialHelper()
+    : anonymous_credential_manager_(
+          anonymous_credentials::new_credential_manager()) {}
+
+BackgroundCredentialHelper::~BackgroundCredentialHelper() = default;
+
+void BackgroundCredentialHelper::UseFixedSeedForTesting() {
+  anonymous_credential_manager_ =
+      anonymous_credentials::new_credential_manager_with_fixed_seed();
+}
+
+std::unique_ptr<RSAKeyInfo> BackgroundCredentialHelper::GenerateRSAKey() {
+  auto key_pair = GenerateRSAKeyPair();
+  if (!key_pair) {
+    return nullptr;
+  }
+  rsa_private_key_ = std::move(key_pair->key_pair);
+  return key_pair;
+}
+
+void BackgroundCredentialHelper::SetRSAKey(
+    std::unique_ptr<crypto::RSAPrivateKey> rsa_private_key) {
+  rsa_private_key_ = std::move(rsa_private_key);
+}
+
+std::optional<GenerateJoinRequestResult>
+BackgroundCredentialHelper::GenerateJoinRequest(std::string pre_challenge) {
+  base::AssertLongCPUWorkAllowed();
+  CHECK(rsa_private_key_);
+  auto challenge = crypto::SHA256Hash(base::as_byte_span(pre_challenge));
+
+  auto join_request = anonymous_credential_manager_->start_join(
```

Comment:
this seems to be changing naming conventions, `start_join` returns a JoinResult, not a JoinRequest